<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to Partition • partition</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to Partition">
<meta property="og:description" content="partition">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">partition</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction-to-partition.html">An Introduction to partition</a>
    </li>
    <li>
      <a href="../articles/extending-partition.html">Extending partition</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/USCbiostats/partition/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to Partition</h1>
                        <h4 class="author">Malcolm Barrett</h4>
            
            <h4 class="date">2020-05-24</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/USCbiostats/partition/blob/master/vignettes/introduction-to-partition.Rmd"><code>vignettes/introduction-to-partition.Rmd</code></a></small>
      <div class="hidden name"><code>introduction-to-partition.Rmd</code></div>

    </div>

    
    
<div id="introduction-to-the-partition-package" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction-to-the-partition-package" class="anchor"></a>Introduction to the partition package</h2>
<p>partition is a fast and flexible data reduction framework for R <span class="citation">(Millstein et al. 2020)</span>. There are many approaches to data reduction, such as principal components analysis (PCA) and hierarchical clustering (both supported in base R). In contrast, partition attempts to create a reduced data set that is both interpretable (each raw feature maps to one and only one reduced feature) and information-rich (reduced features must meet an information constraint). Reducing the data this way often results in a data set that has a mix of raw features from the original data and reduced features.</p>
<p>partition is particularly useful for highly correlated data, such as genomic data, where there is a lot of redundancy. A simple model of say, gene expression data could be block correlated Gaussian variables. <code><a href="../reference/simulate_block_data.html">simulate_block_data()</a></code> simulates data like this: blocks of correlated data that are themselves independent of the other blocks in the data.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">partition</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">1234</span>)
<span class="co"># create a 100 x 15 data set with 3 blocks</span>
<span class="no">df</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/simulate_block_data.html">simulate_block_data</a></span>(
  <span class="co"># create 3 correlated blocks of 5 features each</span>
  <span class="kw">block_sizes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="fl">5</span>, <span class="fl">3</span>),
  <span class="kw">lower_corr</span> <span class="kw">=</span> <span class="fl">.4</span>,
  <span class="kw">upper_corr</span> <span class="kw">=</span> <span class="fl">.6</span>,
  <span class="kw">n</span> <span class="kw">=</span> <span class="fl">100</span>
)</pre></body></html></div>
<p>In a heatmap showing the correlations between the simulated features, blocks of correlated features are visible:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="kw pkg">ggcorrplot</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggcorrplot/man/ggcorrplot.html">ggcorrplot</a></span>(<span class="fu"><a href="../reference/corr.html">corr</a></span>(<span class="no">df</span>))</pre></body></html></div>
<p><img src="introduction-to-partition_files/figure-html/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Many types of data follow a pattern like this. Closely related to the block correlation structure found in genetic data is that found in microbiome data. The data set <code>baxter_otu</code> has microbiome data on 172 healthy patients. Each row represents a patient, and each column represents an Operational Taxonomic Unit (OTU). OTUs are species-like relationships between bacteria determined by analyzing their RNA. Each cell in the dataset represents the logged-count of an OTU found in a patient’s stool sample, with 1,234 OTUs in all.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="no">baxter_otu</span>
<span class="co">#&gt; # A tibble: 172 x 1,234</span>
<span class="co">#&gt;    otu_1 otu_2 otu_3 otu_4 otu_5 otu_6 otu_7 otu_8 otu_9 otu_10 otu_11 otu_12</span>
<span class="co">#&gt;    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt;  1  0     0        0  1.39     0  0        0     0     0  2.08   0       0   </span>
<span class="co">#&gt;  2  2.48  0        0  0        0  0        0     0     0  0      0       0   </span>
<span class="co">#&gt;  3  1.10  0        0  3.26     0  0        0     0     0  0      0       0   </span>
<span class="co">#&gt;  4  1.39  0        0  1.10     0  1.10     0     0     0  0.693  0.693   0   </span>
<span class="co">#&gt;  5  0     0        0  3.00     0  0        0     0     0  0      0       2.40</span>
<span class="co">#&gt;  6  0     0        0  0        0  0        0     0     0  0      0       0   </span>
<span class="co">#&gt;  7  0     0        0  0        0  0        0     0     0  0      0       0   </span>
<span class="co">#&gt;  8  0     0        0  0        0  0        0     0     0  0      0       0   </span>
<span class="co">#&gt;  9  0     0        0  0        0  0        0     0     0  0      0       1.79</span>
<span class="co">#&gt; 10  1.39  1.79     0  0        0  0        0     0     0  0      0       0   </span>
<span class="co">#&gt; # … with 162 more rows, and 1,222 more variables: otu_13 &lt;dbl&gt;, otu_14 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   otu_15 &lt;dbl&gt;, otu_16 &lt;dbl&gt;, otu_17 &lt;dbl&gt;, otu_18 &lt;dbl&gt;, otu_19 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   otu_20 &lt;dbl&gt;, otu_21 &lt;dbl&gt;, otu_22 &lt;dbl&gt;, …</span></pre></body></html></div>
<p>While not as apparent as simulated data, correlated blocks also appear in these data; bacteria tend to group together into communities or cliques in the microbiomes of participants. Here are the first 200 OTUs:</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">correlation_subset</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/corr.html">corr</a></span>(<span class="no">baxter_otu</span>[, <span class="fl">1</span>:<span class="fl">200</span>])
<span class="kw pkg">ggcorrplot</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggcorrplot/man/ggcorrplot.html">ggcorrplot</a></span>(<span class="no">correlation_subset</span>, <span class="kw">hc.order</span> <span class="kw">=</span> <span class="fl">TRUE</span>) + <span class="kw pkg">ggplot2</span><span class="kw ns">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_void</a></span>()</pre></body></html></div>
<p><img src="introduction-to-partition_files/figure-html/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div id="reducing-data-with-partition" class="section level1">
<h1 class="hasAnchor">
<a href="#reducing-data-with-partition" class="anchor"></a>Reducing data with partition</h1>
<p>Because there are many more features (OTUs) in these data than rows (patients), it’s useful to reduce the data for use in statistical modeling. The primary function, <code><a href="../reference/partition.html">partition()</a></code>, takes a data frame and an information threshold and reduces the data to as few variables as possible, subject to the information constraint.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">prt</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/partition.html">partition</a></span>(<span class="no">baxter_otu</span>, <span class="kw">threshold</span> <span class="kw">=</span> <span class="fl">.5</span>)

<span class="no">prt</span>
<span class="co">#&gt; Partitioner:</span>
<span class="co">#&gt;    Director: Minimum Distance (Pearson) </span>
<span class="co">#&gt;    Metric: Intraclass Correlation </span>
<span class="co">#&gt;    Reducer: Scaled Mean</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Reduced Variables:</span>
<span class="co">#&gt; 63 reduced variables created from 158 observed variables</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Mappings:</span>
<span class="co">#&gt; reduced_var_1 = {otu_663, otu_879}</span>
<span class="co">#&gt; reduced_var_2 = {otu_556, otu_662}</span>
<span class="co">#&gt; reduced_var_3 = {otu_269, otu_714}</span>
<span class="co">#&gt; reduced_var_4 = {otu_200, otu_989}</span>
<span class="co">#&gt; reduced_var_5 = {otu_148, otu_704}</span>
<span class="co">#&gt; reduced_var_6 = {otu_75, otu_1111}</span>
<span class="co">#&gt; reduced_var_7 = {otu_793, otu_986, otu_1088}</span>
<span class="co">#&gt; reduced_var_8 = {otu_234, otu_1165}</span>
<span class="co">#&gt; reduced_var_9 = {otu_160, otu_636, otu_830}</span>
<span class="co">#&gt; reduced_var_10 = {otu_232, otu_1141}</span>
<span class="co">#&gt; ...with 53 more reduced variables</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Minimum information:</span>
<span class="co">#&gt; 0.5</span></pre></body></html></div>
<p>For the microbiome data, <code><a href="../reference/partition.html">partition()</a></code> reduced 158 of the OTUs to 63 reduced features. The other 1,076 OTUs were not reduced because doing so would have removed too much information from the data. <code><a href="../reference/partition.html">partition()</a></code> creates a <code>tibble</code> with the newly reduced features, as well as any features that were not reduced, which you can get with <code><a href="../reference/partition_scores.html">partition_scores()</a></code>:</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="fu"><a href="../reference/partition_scores.html">partition_scores</a></span>(<span class="no">prt</span>)
<span class="co">#&gt; # A tibble: 172 x 1,139</span>
<span class="co">#&gt;    otu_1 otu_2 otu_3 otu_4 otu_5 otu_6 otu_7 otu_8 otu_9 otu_10 otu_11 otu_12</span>
<span class="co">#&gt;    &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt;  1  0     0        0  1.39     0  0        0     0     0  2.08   0       0   </span>
<span class="co">#&gt;  2  2.48  0        0  0        0  0        0     0     0  0      0       0   </span>
<span class="co">#&gt;  3  1.10  0        0  3.26     0  0        0     0     0  0      0       0   </span>
<span class="co">#&gt;  4  1.39  0        0  1.10     0  1.10     0     0     0  0.693  0.693   0   </span>
<span class="co">#&gt;  5  0     0        0  3.00     0  0        0     0     0  0      0       2.40</span>
<span class="co">#&gt;  6  0     0        0  0        0  0        0     0     0  0      0       0   </span>
<span class="co">#&gt;  7  0     0        0  0        0  0        0     0     0  0      0       0   </span>
<span class="co">#&gt;  8  0     0        0  0        0  0        0     0     0  0      0       0   </span>
<span class="co">#&gt;  9  0     0        0  0        0  0        0     0     0  0      0       1.79</span>
<span class="co">#&gt; 10  1.39  1.79     0  0        0  0        0     0     0  0      0       0   </span>
<span class="co">#&gt; # … with 162 more rows, and 1,127 more variables: otu_13 &lt;dbl&gt;, otu_14 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   otu_15 &lt;dbl&gt;, otu_16 &lt;dbl&gt;, otu_17 &lt;dbl&gt;, otu_19 &lt;dbl&gt;, otu_20 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   otu_21 &lt;dbl&gt;, otu_22 &lt;dbl&gt;, otu_23 &lt;dbl&gt;, …</span></pre></body></html></div>
<p>In comparison, PCA produces a data set of the same dimensions of the original data where all original variables map to all new components in some amount. While components are organized by their informativeness (the first component explains the most variance), no original features are retained.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">pca</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span>(<span class="no">baxter_otu</span>)

<span class="co"># print the results more neatly</span>
<span class="kw pkg">tibble</span><span class="kw ns">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span>(<span class="no">pca</span>$<span class="no">x</span>)
<span class="co">#&gt; # A tibble: 172 x 172</span>
<span class="co">#&gt;       PC1    PC2      PC3    PC4   PC5    PC6     PC7    PC8   PC9   PC10</span>
<span class="co">#&gt;     &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt;  1  -7.52  5.45   -1.43    1.38  -3.75  6.94  -6.89    4.95  -1.80  2.66 </span>
<span class="co">#&gt;  2   1.62 -4.42   14.9     2.40   3.95  4.30   3.63   -5.84  -1.31  2.78 </span>
<span class="co">#&gt;  3  -9.97 -5.05  -10.2    -0.628  6.36  3.79   0.614  -1.17   1.68  1.28 </span>
<span class="co">#&gt;  4 -10.6   3.91   -1.78   -2.91   1.03  7.24   0.528   8.40   8.26  2.89 </span>
<span class="co">#&gt;  5 -16.2  -4.19    2.76    9.84   1.79  1.62  -6.23   -0.543  4.11  1.63 </span>
<span class="co">#&gt;  6  14.6   0.339  -0.0895  4.11  -1.61 -7.36   5.92    0.735 -6.61  2.08 </span>
<span class="co">#&gt;  7   4.84 -5.08    3.34   -2.74  -2.57 -2.13   6.25    0.858 -1.22 -2.78 </span>
<span class="co">#&gt;  8 -14.2  -7.79    5.29    8.99  -2.38 -0.718  0.0578 -0.676 -2.19  3.13 </span>
<span class="co">#&gt;  9  -5.47 -3.48   -1.86   -4.15  -2.62 -0.510  4.94   -9.12   5.12 -0.126</span>
<span class="co">#&gt; 10 -11.8   9.68   -0.394   4.92  -1.20 -4.57  -6.54   -2.72   6.81  0.181</span>
<span class="co">#&gt; # … with 162 more rows, and 162 more variables: PC11 &lt;dbl&gt;, PC12 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   PC13 &lt;dbl&gt;, PC14 &lt;dbl&gt;, PC15 &lt;dbl&gt;, PC16 &lt;dbl&gt;, PC17 &lt;dbl&gt;, PC18 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   PC19 &lt;dbl&gt;, PC20 &lt;dbl&gt;, …</span></pre></body></html></div>
<p>Notably, these approaches can be easily combined (see below).</p>
</div>
<div id="the-partition-algorithm" class="section level1">
<h1 class="hasAnchor">
<a href="#the-partition-algorithm" class="anchor"></a>The partition algorithm</h1>
<p>partition uses an approach called Direct-Measure-Reduce to create agglomerative (bottom-up) partitions that capture the user-specified minimum level of information. Each variable starts as an individual partition subset, and candidate partition subsets are assessed by verifying that the minimum information is captured in the reduced variable. Reduced variables are easily interpretable because original variables map to one and only one variable in the reduced data set. The partition software is flexible and customizable in the way features are agglomerated, information is measured, and data are reduced.</p>
<p>In this partition, 63 reduced features consist of two to seven features each, as well as 1,076 of the original features that did not get reduced because reducing them would lose too much information. Here are the top 20 clusters, ordered by how many raw features they represent:</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="../reference/plot_partitions.html">plot_ncluster</a></span>(<span class="no">prt</span>, <span class="kw">show_n</span> <span class="kw">=</span> <span class="fl">20</span>) +
  <span class="co"># plot_*() functions return ggplots, so they can be extended using ggplot2</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>(<span class="fl">14</span>)</pre></body></html></div>
<p><img src="introduction-to-partition_files/figure-html/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Each reduced feature explains at least 50% of the information of the original features that it summarizes. The distribution of information has a lower limit of our threshold, .5.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu"><a href="../reference/plot_partitions.html">plot_information</a></span>(<span class="no">prt</span>, <span class="kw">geom</span> <span class="kw">=</span> <span class="no">geom_histogram</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>(<span class="fl">14</span>)</pre></body></html></div>
<p><img src="introduction-to-partition_files/figure-html/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Retrieve a key for these mappings and the information each feature explains with <code><a href="../reference/mapping_key.html">mapping_key()</a></code>, which returns a nested <code>tibble</code>.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu"><a href="../reference/mapping_key.html">mapping_key</a></span>(<span class="no">prt</span>)
<span class="co">#&gt; # A tibble: 1,139 x 4</span>
<span class="co">#&gt;    variable mapping   information indices  </span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;list&gt;          &lt;dbl&gt; &lt;list&gt;   </span>
<span class="co">#&gt;  1 otu_1    &lt;chr [1]&gt;           1 &lt;int [1]&gt;</span>
<span class="co">#&gt;  2 otu_2    &lt;chr [1]&gt;           1 &lt;int [1]&gt;</span>
<span class="co">#&gt;  3 otu_3    &lt;chr [1]&gt;           1 &lt;int [1]&gt;</span>
<span class="co">#&gt;  4 otu_4    &lt;chr [1]&gt;           1 &lt;int [1]&gt;</span>
<span class="co">#&gt;  5 otu_5    &lt;chr [1]&gt;           1 &lt;int [1]&gt;</span>
<span class="co">#&gt;  6 otu_6    &lt;chr [1]&gt;           1 &lt;int [1]&gt;</span>
<span class="co">#&gt;  7 otu_7    &lt;chr [1]&gt;           1 &lt;int [1]&gt;</span>
<span class="co">#&gt;  8 otu_8    &lt;chr [1]&gt;           1 &lt;int [1]&gt;</span>
<span class="co">#&gt;  9 otu_9    &lt;chr [1]&gt;           1 &lt;int [1]&gt;</span>
<span class="co">#&gt; 10 otu_10   &lt;chr [1]&gt;           1 &lt;int [1]&gt;</span>
<span class="co">#&gt; # … with 1,129 more rows</span></pre></body></html></div>
<p>To see each mapping, unnest them using <code><a href="../reference/mapping_key.html">unnest_mappings()</a></code> (or do it yourself with <code><a href="https://tidyr.tidyverse.org/reference/nest.html">tidyr::unnest()</a></code>)</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu"><a href="../reference/mapping_key.html">unnest_mappings</a></span>(<span class="no">prt</span>)
<span class="co">#&gt; # A tibble: 1,234 x 4</span>
<span class="co">#&gt;    variable mapping information indices</span>
<span class="co">#&gt;    &lt;chr&gt;    &lt;chr&gt;         &lt;dbl&gt;   &lt;int&gt;</span>
<span class="co">#&gt;  1 otu_1    otu_1             1       1</span>
<span class="co">#&gt;  2 otu_2    otu_2             1       2</span>
<span class="co">#&gt;  3 otu_3    otu_3             1       3</span>
<span class="co">#&gt;  4 otu_4    otu_4             1       4</span>
<span class="co">#&gt;  5 otu_5    otu_5             1       5</span>
<span class="co">#&gt;  6 otu_6    otu_6             1       6</span>
<span class="co">#&gt;  7 otu_7    otu_7             1       7</span>
<span class="co">#&gt;  8 otu_8    otu_8             1       8</span>
<span class="co">#&gt;  9 otu_9    otu_9             1       9</span>
<span class="co">#&gt; 10 otu_10   otu_10            1      10</span>
<span class="co">#&gt; # … with 1,224 more rows</span></pre></body></html></div>
<div id="direct-measure-reduce-and-partitioners" class="section level2">
<h2 class="hasAnchor">
<a href="#direct-measure-reduce-and-partitioners" class="anchor"></a>Direct-Measure-Reduce and Partitioners</h2>
<p>Partitioners are functions that tell the partition algorithm 1) what to reduce 2) how to measure how much information and 3) how to reduce the data. We call this approach Direct-Measure-Reduce. In partition, functions that handle (1) are thus called directors, functions that handle (2) are called metrics, and functions that handle (3) are called reducers. partition has many pre-specified partitioners, but this aspect of the approach is also quite flexible. See the <a href="extending-partition.html">vignette on extending partition</a> to learn more about custom partitioners.</p>
<p>The default partitioner in <code><a href="../reference/partition.html">partition()</a></code> is <code><a href="../reference/part_icc.html">part_icc()</a></code>. <code><a href="../reference/part_icc.html">part_icc()</a></code> uses a correlation-based distance matrix to find the pair of features with the smallest distance between them; intraclass correlation (ICC) to measure information explained by the reduced feature; and scaled row means to reduce features with a sufficient minimum ICC. <code><a href="../reference/part_icc.html">part_icc()</a></code> is generally fast and scalable.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="../reference/part_icc.html">part_icc</a></span>()
<span class="co">#&gt;    Director: Minimum Distance (Pearson) </span>
<span class="co">#&gt;    Metric: Intraclass Correlation </span>
<span class="co">#&gt;    Reducer: Scaled Mean</span></pre></body></html></div>
<p>There are several other partitioners, which all have names in the format <code>part_*()</code></p>
<table class="table">
<colgroup>
<col width="25%">
<col width="25%">
<col width="25%">
<col width="25%">
</colgroup>
<thead><tr class="header">
<th align="left">partitioner</th>
<th align="left">direct</th>
<th align="left">measure</th>
<th align="left">reduce</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><code><a href="../reference/part_icc.html">part_icc()</a></code></td>
<td align="left">Minimum Distance</td>
<td align="left">ICC</td>
<td align="left">scaled row means</td>
</tr>
<tr class="even">
<td align="left"><code><a href="../reference/part_kmeans.html">part_kmeans()</a></code></td>
<td align="left">K-Means Clusters</td>
<td align="left">Minimum ICC</td>
<td align="left">scaled row means</td>
</tr>
<tr class="odd">
<td align="left"><code><a href="../reference/part_minr2.html">part_minr2()</a></code></td>
<td align="left">Minimum Distance</td>
<td align="left">Minimum R-Squared</td>
<td align="left">scaled row means</td>
</tr>
<tr class="even">
<td align="left"><code><a href="../reference/part_pc1.html">part_pc1()</a></code></td>
<td align="left">Minimum Distance</td>
<td align="left">Variance Explained (PCA)</td>
<td align="left">first principal component</td>
</tr>
<tr class="odd">
<td align="left"><code><a href="../reference/part_icc.html">part_icc()</a></code></td>
<td align="left">Minimum Distance</td>
<td align="left">Standardized Mutual Information</td>
<td align="left">scaled row means</td>
</tr>
</tbody>
</table>
<p>To apply a different partitioner, use the <code>partitioner</code> argument in <code><a href="../reference/partition.html">partition()</a></code>.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">prt_pc1</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/partition.html">partition</a></span>(<span class="no">baxter_otu</span>, <span class="kw">threshold</span> <span class="kw">=</span> <span class="fl">.5</span>, <span class="kw">partitioner</span> <span class="kw">=</span> <span class="fu"><a href="../reference/part_pc1.html">part_pc1</a></span>())
<span class="no">prt_pc1</span>
<span class="co">#&gt; Partitioner:</span>
<span class="co">#&gt;    Director: Minimum Distance (Pearson) </span>
<span class="co">#&gt;    Metric: Variance Explained (PCA) </span>
<span class="co">#&gt;    Reducer: First Principal Component</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Reduced Variables:</span>
<span class="co">#&gt; 370 reduced variables created from 1203 observed variables</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Mappings:</span>
<span class="co">#&gt; reduced_var_1 = {otu_299, otu_720}</span>
<span class="co">#&gt; reduced_var_2 = {otu_301, otu_745}</span>
<span class="co">#&gt; reduced_var_3 = {otu_776, otu_1000}</span>
<span class="co">#&gt; reduced_var_4 = {otu_636, otu_992}</span>
<span class="co">#&gt; reduced_var_5 = {otu_545, otu_675, otu_1164}</span>
<span class="co">#&gt; reduced_var_6 = {otu_308, otu_967}</span>
<span class="co">#&gt; reduced_var_7 = {otu_266, otu_518}</span>
<span class="co">#&gt; reduced_var_8 = {otu_620, otu_819}</span>
<span class="co">#&gt; reduced_var_9 = {otu_255, otu_646, otu_1163}</span>
<span class="co">#&gt; reduced_var_10 = {otu_290, otu_387, otu_560, otu_993, otu_1045}</span>
<span class="co">#&gt; ...with 360 more reduced variables</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Minimum information:</span>
<span class="co">#&gt; 0.501</span></pre></body></html></div>
</div>
<div id="permutation-tests" class="section level2">
<h2 class="hasAnchor">
<a href="#permutation-tests" class="anchor"></a>Permutation tests</h2>
<p>Data sets with statistically independent variables reduce only at lower thresholds because variables can only be reduced if a substantial proportion of information is discarded. In this set of 10 independent variables, setting the threshold to 0.5 results in a partition with no reduction. <code><a href="../reference/partition.html">partition()</a></code> returns the original data.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="co"># create a data.frame of 10 independent features</span>
<span class="no">ind_df</span> <span class="kw">&lt;-</span> <span class="kw pkg">purrr</span><span class="kw ns">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_dfc</a></span>(<span class="fl">1</span>:<span class="fl">10</span>, ~<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="fl">30</span>))
<span class="no">ind_part</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/partition.html">partition</a></span>(<span class="no">ind_df</span>, <span class="fl">.5</span>)
<span class="no">ind_part</span>
<span class="co">#&gt; Partitioner:</span>
<span class="co">#&gt;    Director: Minimum Distance (Pearson) </span>
<span class="co">#&gt;    Metric: Intraclass Correlation </span>
<span class="co">#&gt;    Reducer: Scaled Mean</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Reduced Variables:</span>
<span class="co">#&gt; 0 reduced variables created from 0 observed variables</span>

<span class="fu"><a href="https://rdrr.io/r/base/identical.html">identical</a></span>(<span class="no">ind_df</span>, <span class="fu"><a href="../reference/partition_scores.html">partition_scores</a></span>(<span class="no">ind_part</span>))
<span class="co">#&gt; [1] TRUE</span></pre></body></html></div>
<p>Comparing partitioning that occurs in our observed data to what occurs in data where all features are statistically independent gives us insight into the structure of dependencies in our data. We can make this type of comparison with the function <code><a href="../reference/plot_partitions.html">plot_stacked_area_clusters()</a></code>, which creates partitions for a series of information thresholds using both the observed data and a permuted (independent) version of the data. In the permuted version, each variable is randomly permuted relative to all other variables, thereby enforcing statistical independence. In general, there are many fewer reduced features for the real data than the independent data due to these dependencies; there is more common information across features in the real data.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu"><a href="../reference/plot_partitions.html">plot_stacked_area_clusters</a></span>(<span class="no">df</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>(<span class="fl">14</span>)</pre></body></html></div>
<p><img src="introduction-to-partition_files/figure-html/unnamed-chunk-15-1.png" width="672" style="display: block; margin: auto;"></p>
<p>partition also has a set of tools for more extensive permutation tests. <code><a href="../reference/map_partition.html">map_partition()</a></code> will fit partitions for a range of thresholds for the observed data; <code><a href="../reference/test_permutation.html">test_permutation()</a></code> will do the same but also for a set of permuted data sets (100 by default). <code><a href="../reference/plot_permutation.html">plot_permutation()</a></code> visualizes the results, comparing information, number of clusters, or the number of raw features reduced.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">perms</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/test_permutation.html">test_permutation</a></span>(<span class="no">df</span>, <span class="kw">nperm</span> <span class="kw">=</span> <span class="fl">10</span>)
<span class="no">perms</span>
<span class="co">#&gt; # A tibble: 6 x 9</span>
<span class="co">#&gt;   target_info observed_info nclusters nreduced partition perm_observed_i…</span>
<span class="co">#&gt;         &lt;dbl&gt;         &lt;dbl&gt;     &lt;int&gt;    &lt;int&gt; &lt;list&gt;               &lt;dbl&gt;</span>
<span class="co">#&gt; 1         0.1         0.119         1       15 &lt;partiti…            0.110</span>
<span class="co">#&gt; 2         0.2         0.218         2       15 &lt;partiti…            0.228</span>
<span class="co">#&gt; 3         0.3         0.395         3       15 &lt;partiti…            0.935</span>
<span class="co">#&gt; 4         0.4         0.426         3       14 &lt;partiti…            1    </span>
<span class="co">#&gt; 5         0.5         0.501         4       10 &lt;partiti…            1    </span>
<span class="co">#&gt; 6         0.6         1             0        0 &lt;partiti…            1    </span>
<span class="co">#&gt; # … with 3 more variables: perm_nclusters &lt;dbl&gt;, perm_nreduced &lt;dbl&gt;,</span>
<span class="co">#&gt; #   permutation &lt;list&gt;</span></pre></body></html></div>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="fu"><a href="../reference/plot_permutation.html">plot_permutation</a></span>(<span class="no">perms</span>, <span class="kw">.plot</span> <span class="kw">=</span> <span class="st">"nreduced"</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span>(<span class="fl">14</span>)</pre></body></html></div>
<p><img src="introduction-to-partition_files/figure-html/unnamed-chunk-17-1.png" width="672" style="display: block; margin: auto;"></p>
<p><code><a href="../reference/plot_partitions.html">plot_ncluster()</a></code> and <code><a href="../reference/plot_partitions.html">plot_information()</a></code>, in addition to plotting individual partitions, also plot the results of <code><a href="../reference/test_permutation.html">test_permutation()</a></code>.</p>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-R-partition">
<p>Millstein, Joshua, Francesca Battaglin, Malcolm Barrett, Shu Cao, Wu Zhang, Sebastian Stintzing, Volker Heinemann, and Heinz-Josef Lenz. 2020. “Partition: A Surjective Mapping Approach for Dimensionality Reduction.” <em>Bioinformatics</em> 36 (3): 676–81. <a href="https://doi.org/10.1093/bioinformatics/btz661">https://doi.org/10.1093/bioinformatics/btz661</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Joshua Millstein, <a href="https://malco.io">Malcolm Barrett</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
